src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs

```diff
@@ -0,0 +1,555 @@
+                                                   isImplicit: true);
+            }
+
+            return condition;
+        }
+
+        private void VisitConditionalBranch(IOperation condition, ref BasicBlock dest, bool sense)
+        {
+oneMoreTime:
+
+            switch (condition.Kind)
+            {
+                case OperationKind.BinaryOperator:
+                    var binOp = (IBinaryOperation)condition;
+                    bool testBothArgs = sense;
+
+                    switch (binOp.OperatorKind)
+                    {
+                        case BinaryOperatorKind.ConditionalOr:
+                            testBothArgs = !testBothArgs;
+                            // Fall through
+                            goto case BinaryOperatorKind.ConditionalAnd;
+
+                        case BinaryOperatorKind.ConditionalAnd:
+                            if (testBothArgs)
+                            {
+                                // gotoif(a != sense) fallThrough
+                                // gotoif(b == sense) dest
+                                // fallThrough:
+
+                                BasicBlock fallThrough = null;
+
+                                VisitConditionalBranch(binOp.LeftOperand, ref fallThrough, !sense);
+                                VisitConditionalBranch(binOp.RightOperand, ref dest, sense);
+
+                                if (fallThrough != null)
+                                {
+                                    AppendNewBlock(fallThrough);
+                                }
+                            }
+                            else
+                            {
+                                // gotoif(a == sense) labDest
+                                // gotoif(b == sense) labDest
+
+                                VisitConditionalBranch(binOp.LeftOperand, ref dest, sense);
+                                condition = binOp.RightOperand;
+                                goto oneMoreTime;
+                            }
+                            return;
+                    }
+
+                    // none of above. 
+                    // then it is regular binary expression - Or, And, Xor ...
+                    goto default;
+
+                case OperationKind.UnaryOperator:
+                    var unOp = (IUnaryOperation)condition;
+                    if (unOp.OperatorKind == UnaryOperatorKind.Not && unOp.Operand.Type.SpecialType == SpecialType.System_Boolean)
+                    {
+                        sense = !sense;
+                        condition = unOp.Operand;
+                        goto oneMoreTime;
+                    }
+                    goto default;
+
+                default:
+                    condition = Visit(condition);
+                    dest = dest ?? new BasicBlock(BasicBlockKind.Block);
+                    LinkBlocks(CurrentBasicBlock, (condition, sense, dest));
+                    _currentBasicBlock = null;
```

Comment:
Assuming the other places that set `_currentBasicBlock` to `null` can be removed, as I argued above. This would be the last one (it doesn't call `AppendNewBlock` right after). 

Then maybe this should instead just append a new block (`AppendNewBlock(new BasicBlock(BasicBlockKind.Block));`, logic that is currently in `CurrentBasicBlock`).
Then `_currentBasicBlock` would always be the same as `_blocks.Last()`, which is conceptually simpler.

Reply:
> Assuming the other places that set _currentBasicBlock to null can be removed, as I argued above. 

I prefer explicitly clear the current block when I am done with it rather than relying on side-effects of other APIs that might or might not be called.


---
In reply to: [162158087](https://github.com/dotnet/roslyn/pull/24263#discussion_r162158087) [](ancestors = 162158087)